src/webgpu/api/validation/createRenderPipeline.spec.ts

```diff
@@ -101,97 +101,122 @@ class F extends ValidationTest {
       sampleCount,
     });
   }
+
+  doCreateRenderPipelineTest(
+    isAsync: boolean,
+    _success: boolean,
+    descriptor: GPURenderPipelineDescriptor,
+    callback: (arg0: GPURenderPipeline) => void = _ => {
+      // dummy default callback
+      _ === _;
+    }
+  ) {
+    if (isAsync) {
+      if (_success) {
+        this.shouldResolve(this.device.createRenderPipelineAsync(descriptor).then(callback));
+      } else {
+        // If creating render pipeline is expected to fail, we just skip the callback
+        this.shouldReject('OperationError', this.device.createRenderPipelineAsync(descriptor));
+      }
+    } else {
+      if (_success) {
+        const pipeline = this.device.createRenderPipeline(descriptor);
+        callback(pipeline);
+      } else {
+        this.expectValidationError(() => {
+          this.device.createRenderPipeline(descriptor);
+        });
+      }
+    }
+  }
 }
 
 export const g = makeTestGroup(F);
 
-g.test('basic_use_of_createRenderPipeline').fn(t => {
-  const descriptor = t.getDescriptor();
-
-  t.device.createRenderPipeline(descriptor);
-});
+g.test('basic_use_of_createRenderPipeline')
+  .params(u => u.combine('isAsync', [false, true]))
+  .fn(async t => {
+    const { isAsync } = t.params;
+    const descriptor = t.getDescriptor();
 
-g.test('at_least_one_color_state_is_required').fn(async t => {
-  const goodDescriptor = t.getDescriptor({
-    targets: [{ format: 'rgba8unorm' }],
+    t.doCreateRenderPipelineTest(isAsync, true, descriptor);
   });
 
-  // Control case
-  t.device.createRenderPipeline(goodDescriptor);
+g.test('at_least_one_color_state_is_required')
+  .params(u => u.combine('isAsync', [false, true]))
+  .fn(async t => {
+    const { isAsync } = t.params;
 
-  // Fail because lack of color states
-  const badDescriptor = t.getDescriptor({
-    targets: [],
-  });
+    const goodDescriptor = t.getDescriptor({
+      targets: [{ format: 'rgba8unorm' }],
+    });
+
+    // Control case
+    t.doCreateRenderPipelineTest(isAsync, true, goodDescriptor);
 
-  t.expectValidationError(() => {
-    t.device.createRenderPipeline(badDescriptor);
+    // Fail because lack of color states
+    const badDescriptor = t.getDescriptor({
+      targets: [],
+    });
+
+    t.doCreateRenderPipelineTest(isAsync, false, badDescriptor);
   });
-});
 
 g.test('color_formats_must_be_renderable')
-  .params(u => u.combine('format', kTextureFormats))
+  .params(u =>
+    u.combine('isAsync', [false, true]).combine('format', kTextureFormats as GPUTextureFormat[])
```

Comment:
Why is this cast necessary? `kTextureFormats` should already have the right type, I think.